### PR TITLE
Make readlink stable (enabled by default) now, with the protocol part gated on version support.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #2518

### CODE REVIEW

File: mirrord/agent/src/file.rs [hunk 1]
```diff
@@ -412,7 +412,7 @@ impl FileManager {
     }
 
     /// Handles our `readlink_detour` with [`std::fs::read_link`].
-    #[tracing::instrument(level = "trace", skip_all)]
+    #[tracing::instrument(level = Level::TRACE, skip_all, err(level = Level::ERROR))]
```
Comment (Razz4780) [inline mirrord/agent/src/file.rs:415]:
```suggestion
    #[tracing::instrument(level = Level::TRACE, skip_all)]
```
This should not be logged globally as an agent error, user might simply use a non-existent path

File: mirrord/config/src/lib.rs [hunk 2]
```diff
@@ -524,6 +524,14 @@ impl LayerConfig {
         self.feature.network.dns.verify(context)?;
         self.feature.network.outgoing.verify(context)?;
 
+        if self.experimental.readlink {
+            context.add_warning(
+                "experimenta.readlink config has been deprecated, and `readlink` is now\
```
Comment (Razz4780) [inline mirrord/config/src/lib.rs:529]:
```suggestion
                "experimental.readlink config has been deprecated, and `readlink` is now\
```

File: mirrord/layer/tests/common/mod.rs [hunk 3]
```diff
@@ -741,6 +741,7 @@ pub enum Application {
     RustRecvFrom,
     RustListenPorts,
     Fork,
+    /// `mirrord/layer/tests/apps/readlink/readlink.c`
```
Comment (Razz4780) [inline mirrord/layer/tests/common/mod.rs:744]: :eyes:

**Review by Razz4780 — CHANGES_REQUESTED** [review]
Small stuff only

File: mirrord/intproxy/src/proxies/simple.rs [hunk 4]
```diff
@@ -217,8 +217,34 @@ impl BackgroundTask for SimpleProxy {
...
+                SimpleProxyMessage::FileReq(
+                    message_id,
+                    layer_id,
+                    req @ FileRequest::ReadLink(_),
+                ) => {
+                    if protocol_version
+                        .as_ref()
+                        .is_some_and(|version| READDIR_BATCH_VERSION.matches(version))
+                    {
+                        self.file_reqs.insert(message_id, layer_id);
```
Comment (aviramha) [inline mirrord/intproxy/src/proxies/simple.rs:230]: This means that if user calls readlink, they'll just crash - that's not what we want

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord/layer/src/file/ops.rs [hunk 5]
```diff
@@ -295,24 +295,18 @@ pub(crate) fn pread(local_fd: RawFd, buffer_size: u64, offset: u64) -> Detour<Re
...
 
 /// Resolves the symbolic link `path`.
-///
-/// Bypassed if the `experimental.readlink` config is not set to `true`.
-#[mirrord_layer_macro::instrument(level = "trace", ret)]
+#[mirrord_layer_macro::instrument(level = Level::TRACE, ret)]
 pub(crate) fn read_link(path: Detour<PathBuf>) -> Detour<ReadLinkFileResponse> {
-    if crate::setup().experimental().readlink {
-        let path = remap_path!(path?);
+    let path = remap_path!(path?);
```
Comment (aviramha) [inline mirrord/layer/src/file/ops.rs:300]: We need here to replace the check of the flag with `if version > BATCHED..` then do logic if not None? (like existing) so it'll call bpass

File: mirrord/intproxy/src/proxies/simple.rs (hunk 4 above)
Comment (meowjesty) [reply mirrord/intproxy/src/proxies/simple.rs:230]: My bad. I changed it to `Bypass` on `NotImplemented` error. This way we avoid having to check the version in the layer.

File: mirrord/layer/tests/common/mod.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/layer/tests/common/mod.rs:744]: Just thought it would be nice to have the test files there. Should I remove or may I keep it?

**Review by aviramha — APPROVED** [review]
Approving to remove blocker - fixes looks ok but haven't looked at the whole thing.

File: mirrord/layer/tests/common/mod.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/layer/tests/common/mod.rs]: Nah, at that point I've already done my `gr` command to see which file it is.

**Review by Razz4780 — APPROVED** [review]

**Pull request merged**